### DISCUSSION TITLE
ZFS send receive on Windows

### DISCUSSION THREAD

**Discussion by derritter88** [openzfsonwindows/openzfs#340] [Q&A]
Anyone tried send/receive where Linux would be the sender and Windows would be the receiver?

**Comment by guenther-alka** [comment]
I have not tried but this is on my todo list for my napp-it web-gui port from Solaris to Windows too so I am interested in this item. Basically there are two options for a send/ receive over network. You can either use ssh or netcat (ncat) for transport. Both are available on Windows.

https://nmap.org/ncat/

**Reply by sskras** [reply]
> use ssh or netcat (ncat) for transport. Both are available on Windows.

Well, different builds of `ssh` for Windows exists:

- a w10-native build,
  `%SystemRoot%\System32\OpenSSH\ssh.exe`
- a Cygwin build,
  `%CYGWIN_ROOT%\bin\ssh.exe`
- an MSYS2 build,
  `%MSYS2_ROOT%\usr\bin\ssh.exe`
- a [Midipix](https://news.ycombinator.com/item?id=9651011) build,
  `%MPIX_ROOT%\bin\ssh`

... and then more.

So I would put that not as "are available on Windows" but rather "have builds for Windows".
IOW, the operator cannot count on `ssh` (and especially `ncat`) to be present out of the box.

Just my .02€ :)

**Reply by guenther-alka** [reply]
I usually prefer netcat (on a secure LAN) as this is simple to use and much faster than ssh. 
I will add it as a binary within my napp-it web gui (that is copy and run) so no need to install anything.

**Reply by sskras** [reply]
@guenther-alka, from what I said I think the same is valid to `netcat`, `ncat`, or `nc`: 

a lot of builds exists, and none is guaranteed to be present on every build of Windows.

**Comment by guenther-alka** [comment]
This is not relevant. Windows does not come with any of the three main netcat options available on Windows. This is different to BSD, Illumos, Linux, or OSX where netcat is part of distributions and you must only care of a different syntax per OS especially in the listen part.

On Windows it is enough to have one tested and working option that you can either download or include ex as a ZFS setup option or in my napp-it especially as netcat is a simple portable copy and run single binary tool to make replication any OS to any OS over LAN possible. This is different to SSH that requires a complicated installation.

see https://github.com/openzfsonwindows/openzfs/discussions/402
for current state of a nearly working replication over netcat on Windows

**Comment by Lipown** [comment]
Did anyone try over ssh please? I want to send snapshot from my Linux machine using Syncoid :)
Not sure what the syntax will be considering windows is using drive letters

**Comment by guenther-alka** [comment]
zfs send/receive does not use driveletters on Windows but the same pool/dataset naming as Linux, does not matter if you use netcat or ssh for data transmission.

**Comment by Lipown** [comment]
Ah.. anyway I tried syncoid from linux host but it errored out:
```
root@pve:~# syncoid -R --no-stream --no-sync-snap --force-delete NAS_2/Main lipown@192.168.0.60:Backup_PC/Main
lipown@192.168.0.60's password: 
-n
WARN: lzop not available on target ssh:-S /tmp/syncoid-lipown@192.168.0.60-1740667626 lipown@192.168.0.60 - sync will continue without compression.
WARN: mbuffer not available on target ssh:-S /tmp/syncoid-lipown@192.168.0.60-1740667626 lipown@192.168.0.60 - sync will continue without target buffering.
WARN: ZFS resume feature not available on target machine - sync will continue without resume support.
'ps' is not recognized as an internal or external command,
operable program or batch file.
NEWEST SNAPSHOT: autosnap_2025-02-27_01:00:09_daily
NEWEST SNAPSHOT: autosnap_2025-02-27_01:00:09_daily
INFO: --no-stream selected; sending newest full snapshot NAS_2/Main@autosnap_2025-02-27_01:00:09_daily (~ 1051.4 GB) to new target filesystem:
'ps' is not recognized as an internal or external command,
operable program or batch file.
Sudo is disabled on this machine. To enable it, go to the Developer Settings page in the Settings app
mbuffer: error: outputThread: error writing to <stdout> at offset 0x20000: Broken pipe
mbuffer: warning: error during output to <stdout>: Broken pipe
 398KiB 0:00:00 [4.16MiB/s] [>                                                                                                                                                                 ]  0%            
CRITICAL ERROR:  zfs send  'NAS_2/Main'@'autosnap_2025-02-27_01:00:09_daily' | pv -p -t -e -r -b -s 1128948217672 | mbuffer  -q -s 128k -m 16M 2>/dev/null | ssh     -S /tmp/syncoid-lipown@192.168.0.60-1740667626 lipown@192.168.0.60 ' sudo zfs receive   -F '"'"'Backup_PC/Main'"'"'' failed: 1280 at /usr/sbin/syncoid line 492.
```

**Comment by guenther-alka** [comment]
syncoid seems to call Linux tools remotely what cannot work on Windows, a pur zfs send over ssh will work if ssh is installed on Windows (not per default and there are different options). This is why I included a Windows netcat in my napp-it distribution to make replication any to any working out of the box.

**Comment by Lipown** [comment]
Thanks. I need to learn zfs send and receive then. Thank you! :)

**Reply by lundman** [reply]
It is more that your syncoid is trying to run `ps` and `sudo` on Windows - I guess because it doesn't know it is Windows. If you do a send by hand, you can confirm it works. You could run ssh inside cygwin/git4windows etc, and have those commands.

**Comment by guenther-alka** [comment]
You can also try my napp-it cs. This is a zfs web-gui for multiple OpenZFS servers on Free-BSD, Linux, Illumos, OSX, and Windows that can replicate between. It is free for noncommercial use.

**Reply by Lipown** [reply]
I will definitely try this today! Thanks!
